allennlp_rc/models/transformer_qa.py

```diff
@@ -0,0 +1,245 @@
+            possibly be an answer.
+        span_start : ``torch.IntTensor``, optional
+            From an ``IndexField``. This is one of the things we are trying to predict - the beginning position of
+            the answer.  This is an `inclusive` token index. If this is given, we will compute a loss that gets
+            included in the output dictionary.
+        span_end : ``torch.IntTensor``, optional
+            From an ``IndexField``. This is one of the things we are trying to predict - the ending position of the
+            answer with the passage.  This is an `exclusive` token index. If this is given, we will compute a loss
+            that gets included in the output dictionary.
+        metadata : ``List[Dict[str, Any]]``, optional
+            If present, this should contain the question id, and the original texts of context, question, tokenized
+            version of both, and a list of possible answers. The length of the ``metadata`` list should be the
+            batch size, and each dictionary should have the keys ``id``, ``question``, ``context``,
+            ``question_tokens``, ``context_tokens``, and ``answers``.
+
+        Returns
+        -------
+        An output dictionary consisting of:
+        span_start_logits : torch.FloatTensor
+            A tensor of shape ``(batch_size, passage_length)`` representing unnormalized log
+            probabilities of the span start position.
+        span_start_probs : torch.FloatTensor
+            The result of ``softmax(span_start_logits)``.
+        span_end_logits : torch.FloatTensor
+            A tensor of shape ``(batch_size, passage_length)`` representing unnormalized log
+            probabilities of the span end position (inclusive).
+        span_end_probs : torch.FloatTensor
+            The result of ``softmax(span_end_logits)``.
+        best_span : torch.IntTensor
+            The result of a constrained inference over ``span_start_logits`` and
+            ``span_end_logits`` to find the most probable span.  Shape is ``(batch_size, 2)``
+            and each offset is a token index.
+        best_span_scores : torch.FloatTensor
+            The score for each of the best spans.
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+        best_span_str : List[str]
+            If sufficient metadata was provided for the instances in the batch, we also return the
+            string from the original passage that the model thinks is the best answer to the
+            question.
+        """
+        embedded_question = self._text_field_embedder(question)
+        logits = self._linear_layer(embedded_question)
+        span_start_logits, span_end_logits = logits.split(1, dim=-1)
+        span_start_logits = span_start_logits.squeeze(-1)
+        span_end_logits = span_end_logits.squeeze(-1)
+
+        possible_answer_mask = torch.zeros_like(question["tokens"]["mask"])
```

Comment:
If you use a `SpanField` to get these into a single tensor, there's probably a way to do this using advanced indexing and get rid of the for loop.

Reply:
Also, I'd recommend using `util. get_token_ids_from_text_field_tensors` here instead of reaching into the question field.  It'll give you an object of the right shape, which is all you're looking for here.

Reply:
3d8bf49c758801a55a2141fffc807c0dcacc6d8b for the second comment.

For the first, I don't think that's possible. I want to say `possible_answer_mask[XXX] = 1`, but the region that is indexed by `XXX` is not rectangular. The length of the context is different for every instance. In that case, what would `possible_answer_mask[XXX]` return?

Reply:
You can get regions that are not rectangular with advanced indexing.  Don't have time to give you an example right now, but search the code and the internet for "advanced indexing", and ask for more info if you get stuck.

Reply:
I can do it with boolean indexing:
```Python
        token_ids = get_token_ids_from_text_field_tensors(question_with_context)
        indices = torch.arange(token_ids.size(1)).expand_as(token_ids)
        possible_answer_mask = torch.zeros_like(token_ids)
        context_starts = context_span[:, 0].unsqueeze(1)
        context_ends = context_span[:, 1].unsqueeze(1)
        possible_answer_mask[
            (context_starts.expand_as(token_ids) <= indices) & (indices <= context_ends.expand_as(token_ids))
        ] = 1
```
I did not consider it before, but I'm also not sure it's an improvement. I would guess that with the small batch size (8), this is actually slower than the for loop (because of all the function calls). On the other hand, if we think of this as a reference that other people will copy in their models, maybe we should teach this style.

Reply:
This is neater:
```Python
        token_ids = get_token_ids_from_text_field_tensors(question_with_context)
        indices = torch.arange(token_ids.size(1)).expand_as(token_ids)
        context_starts = context_span[:, 0].unsqueeze(1)
        context_ends = context_span[:, 1].unsqueeze(1)
        possible_answer_mask = (
            (context_starts.expand_as(token_ids) <= indices) &
            (indices <= context_ends.expand_as(token_ids))
        ).long().to(token_ids.device)
```

Reply:
I'd take my comment as just a suggestion - typically avoiding for loops is better in these models, but if it doesn't make a difference in practice, I'd vote for the simpler code.

Reply:
I _think_ `arange` + [`scatter_`](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.scatter_) could work here.